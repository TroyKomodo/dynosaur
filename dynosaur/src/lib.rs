#![doc = include_str!("../README.md")]

pub use dynosaur_derive::*;
