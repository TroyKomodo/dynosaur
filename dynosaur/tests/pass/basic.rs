use dynosaur::dynosaur;

#[dynosaur(DynMyTrait)]
trait MyTrait {
    type Item;
    async fn foo(&self) -> Self::Item;
}

fn main() {}
