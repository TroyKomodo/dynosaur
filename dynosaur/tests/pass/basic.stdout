#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;
use dynosaur::dynosaur;

trait MyTrait {
    async fn foo(&self)
    -> i32;
}
trait ErasedMyTrait {
    fn foo<'life0, 'dynosaur>(&'life0 self)
    ->
        ::core::pin::Pin<Box<dyn ::core::future::Future<Output = i32> +
        'dynosaur>>
    where
    'life0: 'dynosaur,
    Self: 'dynosaur;
}
impl<DYNOSAUR: MyTrait> ErasedMyTrait for DYNOSAUR {
    fn foo<'life0, 'dynosaur>(&'life0 self)
        ->
            ::core::pin::Pin<Box<dyn ::core::future::Future<Output = i32> +
            'dynosaur>> where 'life0: 'dynosaur, Self: 'dynosaur {
        Box::pin(<Self as MyTrait>::foo(self))
    }
}
struct DynMyTrait<'dynosaur> {
    ptr: *mut (dyn ErasedMyTrait + 'dynosaur),
    owned: bool,
}

fn main() {}
